### PR TITLE
Live Preview Update Using Post Message

### PR DESCRIPTION
**Pull request by Buffalom**
This PR is a follow-up to #5909.

It allows every preview target to selectively send a message to the iFrame using `window.postMessage()` instead of hard reloading the page.
This enables a headless frontend to just reload the data. This part however needs to be implemented seperately by the developer.

If you think this feature should be explained in the docs (probably somewhere around [here](https://statamic.dev/live-preview#preview-targets)) I will open a PR in that repository respectively. Let me know what you think.

---

Closes https://github.com/statamic/ideas/issues/788.

### CODE REVIEW

**Comment by duncanmcclean** [comment]
FYI: There's a link under the 'Reviewers' section on the right hand side for you to turn this PR into a draft, rather than needing to add it to the PR title 😉

**Comment by jasonvarga** [comment]
Tests have been fixed on the 3.3 branch, if you'd like to merge that back in when you can, please.

**Comment by Buffalom** [comment]
@jasonvarga Just updated the branch. Hope I did so as you expected. Let me know if there is something else you would like me to do.
Looking forward to have this merged.

**Comment by okaufmann** [comment]
Hi @jasonvarga is there anything we can do in this PR, so you could check it again and maybe can merge it? We just merged 3.3 into this feature branch to have the latest test.

**Comment by nosthrillz** [comment]
Please get this in

**Comment by Buffalom** [comment]
@jasonvarga let me know if you would prefer this to target 4.0.

**Comment by jasonvarga** [comment]
Sorry for the delay on this one.

May I ask what you'd use the `post_message_data` for? In what situation is `live-preview-update` not good enough?

File: resources/js/components/live-preview/UpdatesIframe.js [hunk 1]
```diff
@@ -9,28 +9,43 @@ export default {
...
-                return;
-            }
+            if (container.firstChild) {
+                const existingIFrameSource = new URL(container.firstChild.src);
+                const newIFrameSource = new URL(iframe.src);
+
+                existingIFrameSource.searchParams.delete('live-preview');
+                newIFrameSource.searchParams.delete('live-preview');
+
+                const iFrameSourceIsEqual = existingIFrameSource.toString() === newIFrameSource.toString();
```
Comment (jasonvarga) [inline resources/js/components/live-preview/UpdatesIframe.js:19]: @Buffalom In what situation would you find the iframe source URL to be different?

Comment (jasonvarga) [reply resources/js/components/live-preview/UpdatesIframe.js:19]: Ah I think I worked it out. This could happen if you change preview targets.

**Review by jasonvarga — APPROVED** [review]

**Comment by jasonvarga** [comment]
Thanks for the feedback. I've also now changed the message to be an object, with some more data.

```js
window.onmessage = function (e) {
  console.log(e.data);
  // {
  //   name: 'statamic.preview.updated',
  //   token: 'the-live-preview-token',
  //   reference: 'entry::the-entry-id'
  // }
}
```

**Comment by jasonvarga** [comment]
(I've left the branch open since it's on your repo, but you're free to delete it now.)

**Comment by Buffalom** [comment]
That's perfect. Thanks again.

**Pull request merged**